### PR TITLE
Adds catch for empty requests.

### PR DESCRIPTION
**Pull request by saralein**

### CODE REVIEW

File: src/main/java/com/saralein/server/connection/Connection.java [hunk 1]
```diff
@@ -5,5 +5,5 @@
 public interface Connection {
     String read() throws IOException;
     void write(byte[] output) throws IOException;
-    void close() throws IOException;
+    void close();
```
Comment (damonkelley) [inline src/main/java/com/saralein/server/connection/Connection.java:8]: It feels a little weird that `close` doesn't throw an `IOException` anymore.

File: src/main/java/com/saralein/server/request/RequestParser.java [hunk 2]
```diff
@@ -10,11 +10,15 @@ public RequestParser() {
         this.parsedRequest = new HashMap<>();
     }
 
-    public Request parse(String request) {
-        String[] fullRequest = split(request, CRLF);
-        createRequestMap(fullRequest);
+    public Request parse(String request) throws Exception {
+        try {
+            String[] fullRequest = split(request, CRLF);
+            createRequestMap(fullRequest);
```
Comment (damonkelley) [inline src/main/java/com/saralein/server/request/RequestParser.java]: No change is necessary here, but I am always skeptical of side-effect methods. I tend to find it easier to reason about when the method returns the thing it is operating on, or returns a new version of that object. Just a heads up 😸

File: src/test/java/com/saralein/server/request/RequestParserTest.java [hunk 3]
```diff
@@ -41,20 +41,24 @@ public void setUp() {
...
 
-        assertEquals(postRequest.getMethod(), parsedPostRequest.getMethod());
-        assertEquals(postRequest.getUri(), parsedPostRequest.getUri());
-        assertEquals(postRequest.getBody(), parsedPostRequest.getBody());
+            assertEquals(postRequest.getMethod(), parsedPostRequest.getMethod());
+            assertEquals(postRequest.getUri(), parsedPostRequest.getUri());
+            assertEquals(postRequest.getBody(), parsedPostRequest.getBody());
+        } catch (Exception e) {
+            fail("Request failed to parse in Request Parser test.");
+        }
```
Comment (damonkelley) [inline src/test/java/com/saralein/server/request/RequestParserTest.java]: Just curious, what prompted this change?

File: src/test/java/com/saralein/server/router/ServerRouterTest.java [hunk 4]
```diff
@@ -28,17 +28,23 @@ public void setUp() {
...
+            put("method", "GET");
+            put("uri", "/");
+            put("version", "HTTP/1.1");
+        }});
+
+        fileRequest = new Request(new HashMap<String, String>(){{
+            put("method", "GET");
+            put("uri", "/cheetara.jpg");
+            put("version", "HTTP/1.1");
+        }});
```
Comment (zdavy) [inline src/test/java/com/saralein/server/router/ServerRouterTest.java]: Looks like each of these requests are individually only used once, we should just initialize them in the specific test that they're required.

**Comment by saralein** [comment]
You all seem to be pointing out the things that I disliked about my approach.  So, I think maybe I just need a different strategy for dealing with this issue.

Bug aside, in general I feel like the parser needs to be reworked.

File: src/test/java/com/saralein/server/request/RequestParserTest.java (hunk 3 above)
Comment (saralein) [reply src/test/java/com/saralein/server/request/RequestParserTest.java]: Adding the thrown exception from parser.

File: src/test/java/com/saralein/server/request/RequestParserTest.java [hunk 5]
```diff
@@ -3,10 +3,14 @@
...
 import static org.junit.Assert.*;
 import org.junit.Before;
+import org.junit.Rule;
 import org.junit.Test;
+import org.junit.rules.ExpectedException;
 
 public class RequestParserTest {
     private RequestParser requestParser;
+    @Rule
+    public final ExpectedException exception = ExpectedException.none();
```
Comment (zdavy) [inline src/test/java/com/saralein/server/request/RequestParserTest.java:13]: What is this rule thing?

Comment (saralein) [reply src/test/java/com/saralein/server/request/RequestParserTest.java:13]: That was an experiment I tried and forgot to remove.  Will remove it.

File: src/test/java/com/saralein/server/request/RequestParserTest.java [hunk 6]
```diff
@@ -50,10 +60,21 @@ public void returnsRequestWithBody() {
...
+            fail("Failed to parse request with body.");
+        }
+    }
 
-        assertEquals(request.getMethod(), parsedRequest.getMethod());
-        assertEquals(request.getUri(), parsedRequest.getUri());
-        assertEquals(request.getBody(), parsedRequest.getBody());
+    @Test
+    public void throwsErrorForBadRequest() {
+        try {
```
Comment (damonkelley) [inline src/test/java/com/saralein/server/request/RequestParserTest.java]: So you can use the `ExpectedException` rule here, but you can also add an argument to `@Test` which can be pretty convenient.

```java
@Test(expected = Exception.class)
public void throwsErrorForBadRequest() {
    requestParser.parse("");
}

**Review by damonkelley — APPROVED** [review]

**Review by zdavy — APPROVED** [review]

File: src/test/java/com/saralein/server/request/RequestParserTest.java (hunk 6 above)
Comment (saralein) [reply src/test/java/com/saralein/server/request/RequestParserTest.java]: Nifty.  Thank you very much.  I think I saw this somewhere but didn't fully process its purpose at the time.  👍

**Pull request merged**